src/Network/Transport/TCP.hs

```diff
@@ -629,6 +644,64 @@ apiNewEndPoint transport =
     resolveMulticastGroupError =
       TransportError ResolveMulticastGroupUnsupported "Multicast not supported"
 
+-- | Abstraction of a queue for an 'EndPoint'.
+--
+--   This specifies which 'Event's will come from
+--   'receive :: EndPoint -> IO Event' and when. It is highly general so that
+--   the simple yet potentially very fast implementation backed by a single
+--   unbounded channel can be used, without excluding more nuanced policies
+--   like class-based queueing with bounded buffers for each peer, which may be
+--   faster in certain conditions but probably has lower maximal throughput.
+--
+--   A 'QDisc' must satisfy some properties in order for the semantics of
+--   network-transport to hold true. To grok these, it may help to know how
+--   network-transport-tcp will use the 'QDisc'.
+--
+--   Every call to 'receive' is just 'qdiscDequeue' on that 'EndPoint's
+--   'QDisc'. Whenever an event arises from a socket, `qdiscEnqueue` is called
+--   with the relevant metadata in the same thread that reads from the socket.
+--   You can be clever about when to block here, so as to control network
+--   ingress. This applies also to loopback connections (an 'EndPoint' connects
+--   to itself), in which case blocking on the enqueue would only block some
+--   thread in your program rather than some chatty network peer. The 'Event'
+--   which is to be enqueued is given to 'qdiscEnqueue' so that the 'QDisc'
+--   can know about open connections, their identifiers and peer addresses, etc.
+--
+--   See 'newEndPointInternal', which expects a 'forall t . QDisc t', a queue
+--   discipline which does not use any details of the particular thing it's
+--   queueing.
+data QDisc t = QDisc {
+    qdiscDequeue :: IO t
+  , qdiscEnqueue :: Event -> t -> IO ()
```

Comment:
There's two things pulling in opposite directions here:

- the `forall t` idea is to enforce that the queuing does not alter the events as it passes them through
- but it may be useful for certain policies to be able to distinguish different connections or endpoints (e.g. to do weighted choices) and the simplest way to do that is to be able to inspect the event, or some subset of the info in the event. A previous version of the patch supplied a subset of the info (omitting the data payload), but edsko suggested this was an unnecessary abstraction.

I also now realise that for one policy we'd like to implement we will need the `qdiscEnqueue` to know the `EndPointAddress` of the peer that it's receiving from. This is to be able to do weighted choice between different peers. Indeed it's not clear to me that the content of the event is needed (except opaquely to enqueue).